Added Support for TSQL's constraint_table_usage

oscarno22 opened the pull request:
### Description

Currently, Babelfish does not support
information_schema_tsql.constraint_table_usage. This commit updates this
state by implementing the view.

TASK:

Author: Ozzy Nolen <oscarno@amazon.com>
Signed-off by:
 
### Issues Resolved

- The view information_schema_tsql.constraint_table_usage not implemented


### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

oscarno22 committed:
Added Support for TSQL's constraint_table_usage

Currently, Babelfish does not support
information_schema_tsql.constraint_table_usage. This commit updates this
state by implementing the view.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Updated View to TSQL Syntax

This commit updates the view definition to TSQL syntax, especially as
it pertains to CAST(). The verify test file is also reinitialized.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Updated Testing Files

This commit updates testing files to account for functionality and
dependency. ISC constraint_table_usage is called with the creation of
tables and verified with a view, function, and procedure that depend on
it.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Updated Naming in Test Files

This commit updates the view, function, and procedure created in the
verify file to follow the naming convention:
"constraint_table_usage_vu_verify_%".

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

surendravishnoi commented on test/JDBC/expected/constraint_table_usage-vu-prepare.out:
```diff
@@ -0,0 +1,23 @@
+use master;
+go
+
+create database constraint_table_usage_vu_prepare_db;
```
we do not need to create a separate database

oscarno22 replied:
Resolved in latest update.

oscarno22 committed:
Removed Custom Database from Tests

This commit removes the custom test database and opts to use the one
created in the testing framework.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Updated File for Versio Upgrade Tests

This commit updates expected_dependency.out file to account for
information_schema_tsql.constraint_table_usage.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Update Columns to Retrieve Logical Rolename

Previously, the view definition pulled the physical pg role name for
table_schema and constraint_schema. This commit retrieves the logical
role name from sys.babelfish_authid_username_ext.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Updated vu-verify File for Upgrade Tests

This commit updates the vu-verify file by adding an "order by" clause
when calling the view. This will ensure that the output is consistent
with what is expected.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

shalinilohia50 commented on test/JDBC/expected/constraint_table_usage-vu-verify.out:
```diff
@@ -0,0 +1,58 @@
+select * from information_schema.constraint_table_usage where table_name like 'constraint_table_usage_vu_prepare_tb%' order by table_name,constraint_name,table_schema;
```
The user that can access a table should be able to see its data in the view. Please add tests to verify the behaviour is correct in these cases.

Repro:

> User1:
create table test1 (a int primary key)
grant SELECT on test1 to User2
User2:
SELECT * FROM information_schema.CONSTRAINT_TABLE_USAGE
Should get the data here

JimFinnerty replied:
And also check that the user does not see rows corresponding to objects that they do not have privileges to

oscarno22 replied:
The behavior in SQL Server and PG is to show information for all corresponding objects, regardless of privileges. I think it is best to keep the behavior consistent for now.

oscarno22 replied:
We should use "setuser" or"execute as" to specify from which user the view should retrieve data through in a .sql file. The view will return all previous data without this clarification. However, neither are supported in Babelfish currently. In light of this, I have adjusted my "where" condition to include has_table_privilege() and has_any_column_privilege(). This is a widely accepted syntax in information schema views that have already been implemented.

oscarno22 replied:
An additional testing file was added in the latest commit to check the "select" privilege.

oscarno22 committed:
Updated "where" Condition for Privileges

This commit updates the "where" condition to include the functions
has_table_privilege() and has_any_column_privilege(). The view will
return data when these or pg_has_role() return true.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Added Privilege Test File

This commits adds a testing file specifically for the select privilege. A
.mix file was used in order to switch between different user contexts. A
custom database was used so that role objects would not become
duplicated in the master database.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

surendravishnoi commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -717,3 +717,29 @@ CREATE OR REPLACE VIEW information_schema_tsql.routines AS
+/*
+* CONSTRAINT TABLE USAGE
+*/
+CREATE OR REPLACE VIEW information_schema_tsql.constraint_table_usage
```
what release these are part of? this is upgrade script for lama to manatee. If it is for narwhal, we should create a new upgrade script (I am not sure if branch is open for Narwhal dev or not)

shalinilohia50 replied:
We should have the view definition before this.
```
SELECT set_config('search_path', 'sys, '||current_setting('search_path'), false);
```

oscarno22 committed:
Adding Upgrade Test to Schedule

This commit adds BABEL-1683 to the upgrade test schedules, which should
have been pulled from the upstream.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

HarshLunagariya commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -717,3 +717,29 @@ CREATE OR REPLACE VIEW information_schema_tsql.routines AS
+    CAST(sys.db_name() AS sys.nvarchar(128)) AS "CONSTRAINT_CATALOG",
+    CAST((select orig_name from sys.babelfish_namespace_ext where nc.nspname = nspname) AS sys.nvarchar(128)) AS "CONSTRAINT_SCHEMA",
+    CAST(c.conname AS sys.sysname) AS "CONSTRAINT_NAME"
+   FROM pg_constraint c,
+    pg_namespace nc,
```
Won't we need to filter out the output based on the current database? I'm not sure what will be the current behaviour - whether it will show objects from other db as well or not.

oscarno22 replied:
It doesn't retrieve objects across databases. I have provided a test file on this in this latest commit.

oscarno22 committed:
Added Database Test File

This commit adds an additional test file checking that the view does not
retrieve objects across databases.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

HarshLunagariya commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -717,3 +717,29 @@ CREATE OR REPLACE VIEW information_schema_tsql.routines AS
+AS SELECT CAST(sys.db_name() AS sys.nvarchar(128)) AS "TABLE_CATALOG",
+    CAST((select orig_name from sys.babelfish_namespace_ext where nr.nspname = nspname) AS sys.nvarchar(128)) AS "TABLE_SCHEMA",
+    CAST(r.relname AS sys.sysname) AS "TABLE_NAME",
+    CAST(sys.db_name() AS sys.nvarchar(128)) AS "CONSTRAINT_CATALOG",
```
instead of directly printing sys.db_name() - we should add that as condition to filter out the query result.
i.e. babelfish_namespace_ext.dbid = sys.db_id()
Please refer existing information_schema_tsql views for reference.

oscarno22 committed:
Updated View Definition

This commit replaces pg_namespace with sys.pg_namespace_ext joined by
babelfish_namespace_ext. Also, instead of calling sys.db_name()
directly, the query result is now conditioned.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Added schema test to database test file

This commit checks the output of the view when introducing another schema and
its table.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>
